src/Functions/byteSwap.cpp

```diff
@@ -0,0 +1,79 @@
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionUnaryArithmetic.h>
+
+namespace DB
+{
+namespace ErrorCodes
+{
+extern const int NOT_IMPLEMENTED;
+}
+
+namespace
+{
+template <typename T>
+requires std::is_same_v<T, UInt8>
+inline T byteSwap(T x)
+{
+    return x;
+}
+
+template <typename T>
+requires std::is_same_v<T, UInt16>
+inline T byteSwap(T x)
+{
+    return __builtin_bswap16(x);
+}
+
+template <typename T>
+requires std::is_same_v<T, UInt32>
+inline T byteSwap(T x)
+{
+    return __builtin_bswap32(x);
+}
+
+template <typename T>
+requires std::is_same_v<T, UInt64>
+inline T byteSwap(T x)
+{
+    return __builtin_bswap64(x);
+}
+
+template <typename T>
+inline T byteSwap(T)
+{
+    throw Exception(ErrorCodes::NOT_IMPLEMENTED, "byteSwap() is not implemented for {} datatype", demangle(typeid(T).name()));
```

Comment:
Right now for anything greater than or equal to `2^64`, this prints the following error:
```
SELECT byteSwap(18446744073709552000.)

Query id: 7e644036-8f5a-42ac-be5e-ee1f2286c46e


0 rows in set. Elapsed: 0.005 sec. 

Received exception from server (version 23.9.1):
Code: 48. DB::Exception: Received from localhost:9000. DB::Exception: byteSwap() is not implemented for double datatype: While processing byteSwap(18446744073709552000.). (NOT_IMPLEMENTED)
```

I'd love to know if there's a way to get better type names.

Reply:
Here's my attempt at an implementation for UInt128:
```
template <typename T>
requires std::is_same_v<T, UInt128>
inline T byteSwap(T x)
{
    UInt64 lower_half = x & 0xFFFFFFFFFFFFFFFF;
    UInt64 upper_half = (x >> 64) & 0xFFFFFFFFFFFFFFFF;
    UInt64 swapped_lower_half = __builtin_bswap64(lower_half);
    UInt64 swapped_upper_half = __builtin_bswap64(upper_half);
    UInt128 new_upper_half = static_cast<UInt128> (swapped_lower_half) << 64;
    UInt128 new_lower_half = static_cast<UInt128> (swapped_upper_half);
    return new_upper_half | new_lower_half;
}
```
But I'm not able to test the same. As soon as I go over (2^64-1), for some reason, it doesn't reach the implementation. Also notice that the last four digits of the input change (1616 -> 2000.). Any reasons why this could be happening?
```
clickhouse-400817.internal :) SELECT byteSwap(18446744073709551616);

SELECT byteSwap(18446744073709552000.)

Query id: 0743873e-f48b-423f-a160-1b160f5b25c3


0 rows in set. Elapsed: 0.125 sec. 

Received exception from server (version 23.9.1):
Code: 48. DB::Exception: Received from localhost:9000. DB::Exception: byteSwap() is not implemented for double datatype: While processing byteSwap(18446744073709552000.). (NOT_IMPLEMENTED)
```

Reply:
> But I'm not able to test the same. As soon as I go over (2^64-1), for some reason, it doesn't reach the implementation. Also notice that the last four digits of the input change (1616 -> 2000.). Any reasons why this could be happening?

Because the example is interpreted as float:
```sql
SELECT toTypeName(18446744073709551616);

┌─toTypeName(18446744073709552000.)─┐
│ Float64                           │
└───────────────────────────────────┘
```

Try: `SELECT byteSwap(18446744073709551616::UInt128);`

I wonder why byteSwap on UInt64 and UInt128 operates on 2x8, respectively 4x8 bytes instead of 1x16 / 1x32 bytes? To swap all bytes uniformly, you could use `reverseMemcpy()` (base/base/unaligned.h).